File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/CopySample/src/com/hp/workpath/sample/copysample/task/LoadCapabilitiesTask.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.copysample.task;

import android.os.Handler;
import android.os.Looper;

import com.hp.workpath.api.Result;
import com.hp.workpath.api.copier.CopyAttributesCaps;
import com.hp.workpath.sample.copysample.Logger;
import com.hp.workpath.sample.copysample.MainActivity;
import com.hp.workpath.sample.copysample.R;
import com.hp.workpath.sample.copysample.fragments.CopyConfigureFragment;

import java.lang.ref.WeakReference;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Async task to retrieve copy capabilities from printer.
 */
public class LoadCapabilitiesTask {

    private final WeakReference<MainActivity> mContextRef;
    private final WeakReference<CopyConfigureFragment> mFragment;
    private Result result;

    private ExecutorService executor = Executors.newSingleThreadExecutor();
    private Handler handler = new Handler(Looper.getMainLooper());

    public LoadCapabilitiesTask(final MainActivity context, final CopyConfigureFragment fragment) {
        this.mContextRef = new WeakReference<>(context);
        this.mFragment = new WeakReference<>(fragment);
        this.result = new Result();
    }

    public void taskExecute() {
        try {
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    try {
                        onPostExecute(mContextRef.get().requestCaps(mContextRef.get(), result));
                    } catch (Throwable t) {
                        Logger.showResult(null, "CopierService.getCapabilities " + t.getMessage());
                        executor.shutdown();
                        onPostExecute(null);
                    }
                }
            });
        } catch (Exception e) {
            Logger.showResult(null, "CopierService.getCapabilities " + e.getMessage());
            onPostExecute(null);
            executor.shutdown();
        }
    }

    private void onPostExecute(final CopyAttributesCaps caps) {
        handler.post(new Runnable() {
            @Override
            public void run() {
                if (result.getCode() == Result.RESULT_OK) {
                    mFragment.get().loadCapabilities(caps);
                    Logger.showResult(mContextRef.get(), mContextRef.get().getString(R.string.loaded_caps));
                } else {
                    Logger.showResult(mContextRef.get(), "CopierService.getCapabilities", result);
                }
            }
        });
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/CopySample/src/com/hp/workpath/sample/copysample/task/LoadCapabilitiesTask.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.copysample.task

import com.hp.workpath.api.Result
import com.hp.workpath.api.copier.CopyAttributesCaps
import com.hp.workpath.sample.copysample.Logger
import com.hp.workpath.sample.copysample.MainActivity
import com.hp.workpath.sample.copysample.R
import com.hp.workpath.sample.copysample.fragments.CopyConfigureFragment
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.lang.ref.WeakReference

/**
 * Async task to retrieve copy capabilities from printer.
 */
class LoadCapabilitiesTask(context: MainActivity, fragment: CopyConfigureFragment) {
    private val mContextRef: WeakReference<MainActivity> = WeakReference(context)
    private val mFragment: WeakReference<CopyConfigureFragment> = WeakReference(fragment)
    private val result: Result = Result()
    private var mThrowable: Throwable? = null

    suspend fun execute() {
        var caps: CopyAttributesCaps? = null
        try {
            mContextRef.get()?.run {
                caps = requestCaps(this, result)
            }
        } catch (t: Throwable) {
            mThrowable = t
        }
        onPostExecute(caps)
    }

    private suspend fun onPostExecute(caps: CopyAttributesCaps?) {
        withContext(Dispatchers.Main) {
            mContextRef.get()?.run {
                if (caps != null && result.code == Result.RESULT_OK) {
                    mFragment.get()?.loadCapabilities(caps)
                    Logger.showResult(this, getString(R.string.loaded_caps))
                } else if (mThrowable != null) {
                    Logger.showResult(this, "CopierService.getCapabilities ${mThrowable?.message}");
                } else {
                    Logger.showResult(this, "CopierService.getCapabilities", result);
                }
            }
        }
    }
}